src/CoreWCF.Primitives/src/CoreWCF/Security/SecuritySessionServerSettings.cs

```diff
@@ -0,0 +1,2526 @@
+
+        abstract class ServerSecuritySimplexSessionChannel : ServerSecuritySessionChannel
+        {
+            SoapSecurityInputSession session;
+            bool receivedClose;
+            bool canSendCloseResponse;
+            bool sentCloseResponse;
+            RequestContext closeRequestContext;
+            Message closeResponse;
+            // InterruptibleWaitObject inputSessionClosedHandle = new InterruptibleWaitObject(false);
+
+            public ServerSecuritySimplexSessionChannel(
+                SecuritySessionServerSettings settings,
+                //  IServerReliableChannelBinder channelBinder,
+                SecurityContextSecurityToken sessionToken,
+                object listenerSecurityState, SecurityListenerSettingsLifetimeManager settingsLifetimeManager, EndpointAddress address)
+                : base(settings,
+                      //  channelBinder, 
+                      sessionToken, listenerSecurityState, settingsLifetimeManager, address)
+            {
+                this.session = new SoapSecurityInputSession(sessionToken, settings, this);
+            }
+
+            public IInputSession Session
+            {
+                get
+                {
+                    return this.session;
+                }
+            }
+
+            void CleanupPendingCloseState()
+            {
+                lock (ThisLock)
+                {
+                    if (this.closeResponse != null)
+                    {
+                        this.closeResponse.Close();
+                        this.closeResponse = null;
+                    }
+                    if (this.closeRequestContext != null)
+                    {
+                        this.closeRequestContext.Abort();
+                        this.closeRequestContext = null;
+                    }
+                }
+            }
+
+            protected override void AbortCore()
+            {
+                base.AbortCore();
+                this.Settings.RemoveSessionChannel(this.session.Id);
+                CleanupPendingCloseState();
+            }
+
+            protected override void CloseCore(CancellationToken token)
+            {
+                base.CloseCore(token);
+               // this.inputSessionClosedHandle.Abort(this);
+                this.Settings.RemoveSessionChannel(this.session.Id);
+            }
+
+            public Task CloseAsync(CancellationToken token)
+            {
+                OnCloseAsync(token);
+                return Task.CompletedTask;
+            }
+            protected Task OnCloseAsync(CancellationToken token)
+            {
+                // send a close response if one was not sent yet
+                bool wasAborted = SendCloseResponseOnCloseIfRequired(token);
+                if (wasAborted)
+                {
+                    return Task.CompletedTask;
+                }
+                //bool wasInputSessionClosed = this.WaitForInputSessionClose(token, out wasAborted);
```

Comment:
It looks like calling CloseCore is supposed to wait for the channel to be closed before being called. Is there anything in CloseCore that actually needs to wait? If so, we need a mechanism to signal this. The previous way was when calling TryReceiveRequestContext, the inner channel would return a null RequestContext. With our layering being the other way around, if it's needed a possibility it to have the outer dispatcher call DispatchAsync with a null RequestContext which would then set the waiter. If this needs to be reimplemented, make sure it's waited asynchronously.

Reply:
Matt which channel it has to wait for ? I believe now that there no dependency on Channel builder to switch on/off the channel. Once the request is received, the request is processed and close response is sent and request context is set to null. The client is also working fine with this approach. Let's discuss this.